### PR TITLE
Serialization dynamic layer bug

### PR DESCRIPTION
**Pull request by bszmelcz**
Fix serialization for graphs with last dynamic layer not connected to Results. 

Ticket 45199.

### CODE REVIEW

File: inference-engine/src/inference_engine/generic_ie.cpp [hunk 1]
```diff
@@ -152,9 +151,6 @@ void ngraph::op::GenericIE::validate_and_infer_types() {
             set_output_type(output_index, get_precision(output_index), Shape(outputs[output_index].dims));
         }
         initialized++;
-    } else if (reshape) {
```
Comment (ilyachur) [inline inference-engine/src/inference_engine/generic_ie.cpp:155]: Why did you remove this exception?

Comment (bszmelcz) [reply inference-engine/src/inference_engine/generic_ie.cpp:155]: I am testing how to fix the bug,  but from your comment I guess that is not a desired solution

**Comment by jdanieck** [comment]
The check is removed and CI is green. Looks ready to merge.

**Comment by bszmelcz** [comment]
@ilyachur, I tried to implement the changes with disable reshape as we discussed yesterday.  It turned out that in order to use GenericIE in serialization file I need to add plugin_api to cmake. Does it changes something or is it ok to add this dependency?

**Comment by ilyachur** [comment]
Could you please add new tests to cover this bug?

**Comment by jdanieck** [comment]
> Could you please add new tests to cover this bug?

The initial report says it will be issue with NMS. So @bszmelcz let's try to reproduce this issue with NMS by introducing Serialization SLT for it, and then check if your fix is resolving issue.

> Hi, I realized that we merged https://github.com/openvinotoolkit/openvino/pull/3475 but the fix was actually wrong because node->is_dynamic() method checks only input shapes but not output. So it won’t work for NMS which always produce dynamic output shape. I think we need to remove this check at all.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 2]
```diff
@@ -491,10 +491,6 @@ bool is_exec_graph(const ngraph::Function& f) {
 
 bool resolve_dynamic_shapes(const ngraph::Function& f) {
     const auto & f_ops = f.get_ordered_ops();
-    if (std::all_of(f_ops.begin(), f_ops.end(),
-            [](std::shared_ptr<Node> results) { return !results->is_dynamic(); })) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:495]: I refreshed my knowledge with an email you've attached. And yes, you are right that in some cases nodes can always produce dynamic outputs even with static inputs. But removing this if statement we will apply `resolve_dynamic_shapes` for all functions including static. I suggest to check here both inputs and outputs node ports and only if some of them is dynamic then we go next.

**Comment by GlebKazantaev** [comment]
> > Could you please add new tests to cover this bug?
> 
> The initial report says it will be issue with NMS. So @bszmelcz let's try to reproduce this issue with NMS by introducing Serialization SLT for it, and then check if your fix is resolving issue.
> 
> > Hi, I realized that we merged #3475 but the fix was actually wrong because node->is_dynamic() method checks only input shapes but not output. So it won’t work for NMS which always produce dynamic output shape. I think we need to remove this check at all.

I wasn't right about removing this check at all. See my comment above.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 2 above)
Comment (jdanieck) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:495]: Yes, this is better solution. I'll talk tomorrow with @bszmelcz and explain.

**Comment by bszmelcz** [comment]
@GlebKazantaev I added the SLT for serialization for NMS, but it seems like that all the tests are passing. Could you give me some hint what might be wrong, since I cannot reproduce the bug?

File: inference-engine/tests/functional/inference_engine/serialization/single_layer/non_max_suppression.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            ::testing::ValuesIn(maxOutBoxPerClass),
+            ::testing::ValuesIn(threshold), // IOU threshold
+            ::testing::ValuesIn(threshold), // Score threshold
+            ::testing::ValuesIn(sigmaThreshold),
+            ::testing::ValuesIn(encodType),
+            ::testing::ValuesIn(sortResDesc),
+            ::testing::ValuesIn(outType),
+            ::testing::Values(CommonTestUtils::DEVICE_CPU));
+
+    INSTANTIATE_TEST_CASE_P(smoke_NmsLayerTest, NmsLayerTest, nmsParams, NmsLayerTest::getTestCaseName);
```
Comment (ilyachur) [inline inference-engine/tests/functional/inference_engine/serialization/single_layer/non_max_suppression.cpp:56]: Did you check that this test has been failed before?

**Review by ilyachur — APPROVED** [review]

File: inference-engine/tests/functional/inference_engine/serialization/single_layer/non_max_suppression.cpp (hunk 3 above)
Comment (bszmelcz) [reply inference-engine/tests/functional/inference_engine/serialization/single_layer/non_max_suppression.cpp:56]: From email conversation that was attached to the ticket, test for NMS should be failing, but it is passing and besides providing SLT for it I did not make any additional changes.

Comment (jdanieck) [reply inference-engine/tests/functional/inference_engine/serialization/single_layer/non_max_suppression.cpp:56]: @ilyachur current state is that Bartosz can't reproduce this issue with NMS in last layer. I will try to reproduce it myself today, but if I'm not able then we can either reject this PR or add fix without a test to handle 'theoretical' issue.

Comment (jdanieck) [reply inference-engine/tests/functional/inference_engine/serialization/single_layer/non_max_suppression.cpp:56]: @bszmelcz I've reproduced the issue. I've taken NMS-5 as a last layer and connected only it's static output to Result, hence dynamic ones were not connected to anywhere. As a result I got exception during serialization complaining about dynamism.  When added additional checking of op outputs in resolve_dynamic_shapes() it started to pass. I've send you patch on Teams.

Comment (bszmelcz) [reply inference-engine/tests/functional/inference_engine/serialization/single_layer/non_max_suppression.cpp:56]: I was also able to reproduce the issue. I added the fixes that @jdanieck provided and they fix the bug.

**Review by jdanieck — APPROVED** [review]

**Comment by jdanieck** [comment]
@ilyachur I think we can merge, reported issue was reproduced with test and fixed. Additionally we got Serialization SLT for NMS op.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 4]
```diff
@@ -506,10 +506,20 @@ bool is_exec_graph(const ngraph::Function& f) {
     return false;
 }
 
+bool has_dynamic_output(std::shared_ptr<Node> n) {
+  for (size_t i = 0; i < n->get_output_size(); i++) {
+    if (n->get_output_partial_shape(i).is_dynamic()) {
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:511]: nit: incorrect indent

Comment (jdanieck) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:511]: Can we expect .clang-format for IE anytime soon ? If you create such ticket I can happily take it on the next release :)

Comment (ilyachur) [reply inference-engine/src/transformations/src/transformations/serialize.cpp]: @jdanoeck It is a big story. Let's discuss it offline.
